controllers/cache/readiness.go

```diff
@@ -0,0 +1,257 @@
+package cache
+
+import (
+	"errors"
+	"fmt"
+	rediscli "github.com/go-redis/redis"
+	goharborv1 "github.com/goharbor/harbor-cluster-operator/api/v1"
+	"github.com/goharbor/harbor-cluster-operator/controllers/k8s"
+	"github.com/goharbor/harbor-cluster-operator/lcm"
+	corev1 "k8s.io/api/core/v1"
+	kerr "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"strings"
+)
+
+const (
+	HarborChartMuseum = "chartmuseum"
+	HarborClair       = "clair"
+	HarborJobService  = "jobservice"
+	HarborRegistry    = "registry"
+)
+
+var (
+	components = []string{
+		HarborChartMuseum,
+		HarborClair,
+		HarborJobService,
+		HarborRegistry,
+	}
+)
+
+// Readiness reconcile will check Redis sentinel cluster if that has available.
+// It does:
+// - create redis connection pool
+// - ping redis server
+// - return redis properties if redis has available
+func (redis *RedisReconciler) Readiness() error {
+	var (
+		client *rediscli.Client
+		err    error
+	)
+
+	switch redis.HarborCluster.Spec.Redis.Kind {
+	case "external":
+		client, err = redis.GetExternalRedisInfo()
+	case "inCluster":
+		client, err = redis.GetInClusterRedisInfo()
+	}
+
+	if err != nil {
+		redis.Log.Error(err, "Fail to create redis client.", "namespace", redis.Namespace, "name", redis.Name)
+		return err
+	}
+
+	defer client.Close()
+
+	if err := client.Ping().Err(); err != nil {
+		redis.Log.Error(err, "Fail to check Redis.", "namespace", redis.Namespace, "name", redis.Name)
+		return err
+	}
+
+	redis.Log.Info("Redis already ready.", "namespace", redis.Namespace, "name", redis.Name)
+
+	for _, component := range components {
+		url := redis.RedisConnect.GenRedisConnURL()
+		if err := redis.DeployComponentSecret(component, url, ""); err != nil {
+			return err
```

Comment:
What about the successfully created ones previously if here we quit and return an error? They're to be dirty data?